File: element-common/src/main/java/io/github/solusmods/eternalcore/element/impl/network/c2s/SyncElementsStoragePayload.java
package io.github.solusmods.eternalcore.element.impl.network.c2s;

import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import io.github.solusmods.eternalcore.element.EternalCoreElements;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;

public record SyncElementsStoragePayload(CompoundTag data) implements CustomPacketPayload {
    public static final Type<SyncElementsStoragePayload> TYPE = new Type<>(EternalCoreElements.create("sync_elements"));
    public static final StreamCodec<FriendlyByteBuf, SyncElementsStoragePayload> STREAM_CODEC = CustomPacketPayload.codec(SyncElementsStoragePayload::encode, SyncElementsStoragePayload::new);

    public SyncElementsStoragePayload(FriendlyByteBuf buf) {
        this(buf.readNbt());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeNbt(data);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.SERVER) return;
        context.queue(() -> ClientAccess.handle(this, (ServerPlayer) context.getPlayer()));
    }

    /**
     * @return
     */
    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }
}

File: element-common/src/main/kotlin/io/github/solusmods/eternalcore/element/impl/network/c2s/SyncElementsStoragePayload.kt
package io.github.solusmods.eternalcore.element.impl.network.c2s

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.element.EternalCoreElements
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamDecoder
import net.minecraft.network.codec.StreamMemberEncoder
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.server.level.ServerPlayer

@JvmRecord
data class SyncElementsStoragePayload(val data: CompoundTag?) : CustomPacketPayload {
    constructor(buf: FriendlyByteBuf) : this(buf.readNbt())

    fun encode(buf: FriendlyByteBuf) {
        buf.writeNbt(data)
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.environment != Env.SERVER) return
        context.queue { ClientAccess.handle(this, context.player as ServerPlayer) }
    }

    /**
     * @return
     */
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload?> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<SyncElementsStoragePayload?> =
            CustomPacketPayload.Type<SyncElementsStoragePayload?>(EternalCoreElements.create("sync_elements"))
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, SyncElementsStoragePayload?> =
            CustomPacketPayload.codec<FriendlyByteBuf?, SyncElementsStoragePayload?>(
                { obj: SyncElementsStoragePayload?, buf: FriendlyByteBuf? -> obj!!.encode(buf!!) },
                { buf: FriendlyByteBuf? ->
                    SyncElementsStoragePayload(
                        buf!!
                    )
                })
    }
}
